eth/gasprice/scroll_gasprice.go

```diff
@@ -0,0 +1,150 @@
+package gasprice
+
+import (
+	"context"
+	"math/big"
+	"sort"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rpc"
+)
+
+func (oracle *Oracle) CalculateSuggestPriorityFee(ctx context.Context, header *types.Header) *big.Int {
+	// Before Curie (EIP-1559), we need to return the total suggested gas price. After Curie we return defaultGasTipCap wei as the tip cap,
+	// as the base fee is set separately or added manually for legacy transactions.
+	suggestion := oracle.defaultGasTipCap
+	if !oracle.backend.ChainConfig().IsCurie(header.Number) {
+		suggestion = oracle.defaultBasePrice
+	}
+
+	// find the maximum gas used by any of the transactions in the block to use as the gas limit
+	// capacity margin
+	receipts, err := oracle.backend.GetReceipts(ctx, header.Hash())
+	if receipts == nil || err != nil {
+		log.Error("failed to get block receipts", "block number", header.Number, "err", err)
+		return suggestion
+	}
+	var maxTxGasUsed uint64
+
+	for i := range receipts {
+		gu := receipts[i].GasUsed
+		if gu > maxTxGasUsed {
+			maxTxGasUsed = gu
+		}
+	}
+
+	// find the maximum transaction size by any of the transactions in the block to use as the block
+	// size limit capacity margin
+	var (
+		maxTxSizeUsed   common.StorageSize
+		totalTxSizeUsed common.StorageSize
+	)
+	block, err := oracle.backend.BlockByNumber(ctx, rpc.BlockNumber(header.Number.Int64()))
+	if block == nil || err != nil {
+		log.Error("failed to get last block", "err", err)
+		return suggestion
+	}
+	txs := block.Transactions()
+
+	for i := range txs {
+		su := txs[i].Size()
```

Comment:
There's a heavier op `rlp.Encode(&c, &tx.inner)` inside this function. It's nice to have a benchmark test of API QPS in `scroll_gasprice_test.go`, filling blocks with different transactions (since `txs[i].Size()` has a cache).